s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/AivenS3SourceRecord.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2024 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.connect.s3.source.utils;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.kafka.connect.data.SchemaAndValue;
+import org.apache.kafka.connect.source.SourceRecord;
+
+public class AivenS3SourceRecord {
+    private final Map<String, Object> partitionMap;
+    private final Map<String, Object> offsetMap;
+    private final String toTopic;
+    private final int topicPartition;
+    private final byte[] recordKey;
+    private final byte[] recordValue;
+
+    public AivenS3SourceRecord(final Map<String, Object> partitionMap, final Map<String, Object> offsetMap,
+            final String toTopic, final int topicPartition, final byte[] recordKey, final byte[] recordValue) {
+        this.partitionMap = new HashMap<>(partitionMap);
+        this.offsetMap = new HashMap<>(offsetMap);
+
+        this.toTopic = toTopic;
+        this.topicPartition = topicPartition;
+        this.recordKey = Arrays.copyOf(recordKey, recordKey.length);
+        this.recordValue = Arrays.copyOf(recordValue, recordValue.length);
```

Comment:
Just want to point out that Arrays.copyOf is a shallow copy I know below you are doing a clone on the array, but if anything alters the original, it could affect your copy, and vice versa.

Reply:
Indeed, yes, the intent here is to not alter it. so had to put it with Arrays.copyOf.